liblouis/compileTranslationTable.c

```diff
@@ -2618,9 +2617,175 @@ compileBeforeAfter(FileInfo *file) {
+	m->substitutions = substitutions_copy;
+	m->substitution_count = substitution_count;
+	m->argument_count = argument_count;
+	return m;
+}
+
+/**
+ * Create new macro list from "head" macro and "tail" list.
+ */
+static const MacroList *
+cons_macro(const Macro *head, const MacroList *tail) {
+	MacroList *list = malloc(sizeof(MacroList));
+	list->head = head;
+	list->tail = tail;
+	return list;
+}
+
+/**
+ * Free macro returned by create_macro.
+ */
+static void
+free_macro(const Macro *macro) {
+	if (macro) {
+		free((char *)macro->name);
+		free((char *)macro->definition);
+		free((int *)macro->substitutions);
+		free((Macro *)macro);
+	}
+}
+
+/**
+ * Free macro list returned by cons_macro.
+ */
+static void
+free_macro_list(const MacroList *list) {
+	if (list) {
+		free_macro((Macro *)list->head);
+		free_macro_list((MacroList *)list->tail);
+		free((MacroList *)list);
+	}
+}
+
+/**
+ * Compile macro
+ */
+static int
+compileMacro(FileInfo *file, const Macro **macro) {
+
+	// parse name
+	CharsString token;
+	if (!getToken(file, &token, "macro name")) return 0;
+	for (TranslationTableOpcode opcode = 0; opcode < CTO_None; opcode++) {
+		if (token.length == opcodeLengths[opcode] &&
+				eqasc2uni((unsigned char *)opcodeNames[opcode], &token.chars[0],
+						token.length)) {
+			compileError(file, "Invalid macro name: already taken by an opcode");
+			return 0;
+		}
+	}
+	for (int i = 0; i < token.length; i++) {
+		if (!((token.chars[i] >= 'a' && token.chars[i] <= 'z') ||
+					(token.chars[i] >= 'A' && token.chars[i] <= 'Z') ||
+					(token.chars[i] >= '0' && token.chars[i] <= '9'))) {
+			compileError(file,
+					"Invalid macro name: must be a word containing only letters and "
+					"digits");
+			return 0;
+		}
+	}
+	static char name[MAXSTRING + 1];
+	int name_length;
+	for (name_length = 0; name_length < token.length;
+			name_length++)  // we know token can not be longer than MAXSTRING
+		name[name_length] = (char)token.chars[name_length];
+	name[name_length] = '\0';
+
+	// parse body
+	static widechar definition[MAXSTRING];
+	static int substitutions[2 * MAX_MACRO_VAR];
+	int definition_length = 0;
+	int substitution_count = 0;
+	int argument_count = 0;
+	int dollar = 0;
+
+	// ignore rest of line after name and read lines until "eom" is encountered
+	while (_lou_getALine(file)) {
+		if (file->linelen >= 3 && file->line[0] == 'e' && file->line[1] == 'o' &&
+				file->line[2] == 'm') {
+			*macro = create_macro(name, definition, definition_length, substitutions,
+					substitution_count, argument_count);
+			return 1;
+		}
+		while (!atEndOfLine(file)) {
+			widechar c = file->line[file->linepos++];
+			if (dollar) {
+				dollar = 0;
+				if (c >= '0' && c <= '9') {
+					if (substitution_count >= MAX_MACRO_VAR) {
+						compileError(file,
+								"Macro can not have more than %d variable substitutions",
+								MAXSTRING);
+						return 0;
+					}
+					substitutions[2 * substitution_count] = definition_length;
```

Comment:
I find it a bit confusing to have both the number of the substitution and the place where the substitution is supposed to happen in the same array. Could the substitution array not be an array of tuples where you keep arg_number and place? Then there wouldn't be these complicated calculations with `substitutions[2*subst]`